projects/libyaml/libyaml_dumper_fuzzer.c

```diff
@@ -0,0 +1,307 @@
+
+  yaml_node_t *node;
+  yaml_node_item_t *item;
+  yaml_node_pair_t *pair;
+
+  if (!yaml_document_initialize(document_to, document_from->version_directive,
+                                document_from->tag_directives.start,
+                                document_from->tag_directives.end,
+                                document_from->start_implicit,
+                                document_from->end_implicit))
+    return !error;
+
+  for (node = document_from->nodes.start; node < document_from->nodes.top;
+       node++) {
+    switch (node->type) {
+    case YAML_SCALAR_NODE:
+      if (!yaml_document_add_scalar(
+              document_to, node->tag, node->data.scalar.value,
+              node->data.scalar.length, node->data.scalar.style))
+        goto out;
+      break;
+    case YAML_SEQUENCE_NODE:
+      if (!yaml_document_add_sequence(document_to, node->tag,
+                                      node->data.sequence.style))
+        goto out;
+      break;
+    case YAML_MAPPING_NODE:
+      if (!yaml_document_add_mapping(document_to, node->tag,
+                                     node->data.mapping.style))
+        goto out;
+      break;
+    default:
+      goto out;
+    }
+  }
+
+  for (node = document_from->nodes.start; node < document_from->nodes.top;
+       node++) {
+    switch (node->type) {
+    case YAML_SEQUENCE_NODE:
+      for (item = node->data.sequence.items.start;
+           item < node->data.sequence.items.top; item++) {
+        if (!yaml_document_append_sequence_item(
+                document_to, node - document_from->nodes.start + 1, *item))
+          goto out;
+      }
+      break;
+    case YAML_MAPPING_NODE:
+      for (pair = node->data.mapping.pairs.start;
+           pair < node->data.mapping.pairs.top; pair++) {
+        if (!yaml_document_append_mapping_pair(
+                document_to, node - document_from->nodes.start + 1, pair->key,
+                pair->value))
+          goto out;
+      }
+      break;
+    default:
+      break;
+    }
+  }
+  return error;
+
+out:
+  yaml_document_delete(document_to);
+  return !error;
+}
+
+int LLVMFuzzerTestOneInput(const uint8_t *data, size_t size) {
+  if (size < 2)
+    return 0;
+
+  yaml_parser_t parser;
+  yaml_emitter_t emitter;
+
+  yaml_document_t document;
+  unsigned char buffer[BUFFER_SIZE + 1];
+  size_t written = 0;
+  yaml_document_t documents[MAX_DOCUMENTS];
+  size_t document_number = 0;
+  bool done = false;
+  int count = 0;
+  int equal = 0;
+  int k;
+  bool is_canonical = data[0] & 1;
+  bool is_unicode = data[1] & 1;
+  data = data + 2;
+  size = size - 2;
+
+  if (!yaml_parser_initialize(&parser))
+    return 0;
+
+  yaml_parser_set_input_string(&parser, data, size);
+  if (!yaml_emitter_initialize(&emitter))
+    return 0;
+
+  yaml_emitter_set_canonical(&emitter, is_canonical);
+  yaml_emitter_set_unicode(&emitter, is_unicode);
+
+  yaml_emitter_set_output_string(&emitter, buffer, BUFFER_SIZE, &written);
```

Comment:
here you could also use the implementation of the dynamically growing buffer mentioned in a later comment.